src/source/shift.rs

```diff
@@ -4,141 +4,136 @@ use json;
 use binjs_generic::syntax::{ASTError, MutASTVisitor, MutASTWalker, WalkPath };
 
 use source::parser::SourceParser;
 
+use which::which;
+
 #[derive(Debug)]
 pub enum Error {
     CouldNotLaunch(std::io::Error),
-    CouldNotReadFile(std::io::Error),
-    ExecutionError(std::io::Error),
-    CouldNotCreateFile(std::io::Error),
-    ReturnedError(ExitStatus),
-    JsonError(json::JsonError),
+    IOError(std::io::Error),
+    JsonError(json::Error),
     InvalidPath(PathBuf),
-    InvalidUTF8(std::string::FromUtf8Error),
     InvalidAST(ASTError),
+    NodeNotFound(which::Error),
+    ParsingError(String),
 }
 
-/// Using a Node + Shift binary to parse an AST.
-pub struct Shift {
-    bin_path: PathBuf
+pub struct NodeConfig {
+    bin_path: PathBuf,
+    scripts_path: &'static Path,
+    memory: OsString,
 }
 
-impl Shift {
-    pub fn new() -> Self {
-        Shift::with_path("node")
+impl NodeConfig {
+    pub fn try_new() -> Result<Self, Error> {
+        Ok(Self {
+            bin_path: which("node").map_err(Error::NodeNotFound)?,
+
+            scripts_path: Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/src/source")),
+
+            memory: match env::var("NODE_MAX_OLD_SPACE_SIZE") {
+                Err(_) => String::from("--max_old_space_size=2048"),
+                Ok(v) => format!("--max_old_space_size={}", v)
+            }.into(),
+        })
     }
 
-    pub fn with_path<P: AsRef<Path>>(bin_path: P) -> Self {
-        Shift {
-            bin_path: bin_path.as_ref().to_path_buf()
-        }
+    pub fn create_cmd(&self, name: &str) -> Command {
+        let mut cmd = Command::new(&self.bin_path);
+        cmd.env("NODE_PATH", "node_modules");
+        cmd.arg(&self.memory);
+        cmd.arg(self.scripts_path.join(name));
+        cmd
     }
+}
 
-    fn parse_script_output(&self, script: &str) -> Result<String, Error> {
-        debug!(target: "Shift", "Preparing script {}", script);
-
-        let script = format!(r##"
-        var result;
-        try {{
-            result = (function() {{
-                {}
-            }})();
-        }} catch (ex) {{
-            console.warn(ex);
-            /* rethrow */ throw ex;
-        }};
-        var process = require('process');
-        /* See crates/binjs_io/src/escaped_wtf8.rs */
-        result = result
-            .replace(/[\u007F\uD800-\uDFFF]/ug, function(m) {{
-                if (m == "\u007F") {{
-                    return "\u007F007F";
-                }}
-                return "\u007F" + m.charCodeAt(0).toString(16);
-            }});
-        process.stdout.write(result);
-        console.warn(result);
-        "##,
-        script);
-
-        debug!(target: "Shift", "Launching script {}", script);
-
-        let node_memory = match env::var("NODE_MAX_OLD_SPACE_SIZE") {
-            Err(_) => String::from("--max_old_space_size=2048"),
-            Ok(v) => format!("--max_old_space_size={}", v)
-        };
+struct ScriptIO {
+    input: LineWriter<ChildStdin>,
```

Comment:
As usual, doc would be nice :)

Reply:
I don't really know what to add to this one TBH.

I doubt stating "This is lines of input" and "This is lines of output" would be useful given that it's what field names and types already say, so it would be helpful to know what you'd expect here :)